Use digest cache in development.

kaspth opened the pull request:
Fixes #20361

Avoid computing the same fragment digest many times when looping over templates, and
cut down on logging noise.

The cache is cleared on every request, so there's no chance of stale templates being served.

cc @dhh @rafaelfranca @britto

kaspth committed:
Make digest cache work in development.

Avoid computing the same fragment digest many times when looping over templates.

The cache is cleared on every request so template changes are still picked up.

kaspth commented:
Ya, it's kinda late here and I think I missed that we should wrap around `fragment_for` to capture what's written to the cache and not just the fragment name. c/d @dhh?

I should add a test for the caching writing and reading. Possibly even one for the reduced log output.

For glory and all the XP!

dhh commented:
I haven't actually dug into this part of the code in a long time. @rafaelfranca could you vet this?

rafaelfranca commented:
> Ya, it's kinda late here and I think I missed that we should wrap around fragment_for to capture what's written to the cache and not just the fragment name. c/d

c

rafaelfranca commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -42,6 +42,16 @@ class Railtie < Rails::Railtie # :nodoc:
 
+    initializer "action_view.fragment_name_cache" do |app|
+      ActiveSupport.on_load(:action_view) do
+        if Rails.env.development?
```
I think we should check `consider_all_requests_local`.

dhh commented:
Hey @kaspth had any chance to consider this further?

kaspth commented:
Sorry, I've totally sweat this out. I'll take a look this weekend :+1:



--
Kasper

On 26. jun. 2015 15.29 +0200, David Heinemeier Hansson<notifications@github.com>, wrote:
>  
> Hey@kaspth(https://github.com/kaspth)had any chance to consider this further?
>  
>  
> —
> Reply to this email directly orview it on GitHub(https://github.com/rails/rails/pull/20384#issuecomment-115686357).
>  
>  
>

dhh commented:
Rocking. Thanks man!

On Fri, Jun 26, 2015 at 3:45 PM, Kasper Timm Hansen <
notifications@github.com> wrote:

> Sorry, I've totally sweat this out. I'll take a look this weekend :+1:
>
>
>
> --
> Kasper
>
> On 26. jun. 2015 15.29 +0200, David Heinemeier Hansson<
> notifications@github.com>, wrote:
> >
> > Hey@kaspth(https://github.com/kaspth)had any chance to consider this
> further?
> >
> >
> > —
> > Reply to this email directly orview it on GitHub(
> https://github.com/rails/rails/pull/20384#issuecomment-115686357).
>
> >
> >
> >
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20384#issuecomment-115695010>.
>

kaspth commented on actionview/lib/action_view/helpers/cache_helper.rb:
```diff
@@ -98,7 +98,7 @@ module CacheHelper
       #   <%= render_sortable_todolists @project.todolists %>
       #
-      # The pattern used to match these is <tt>/# Template Dependency: (\S+)/</tt>,
+      # The pattern used to match these is /# Template Dependency: ([^ ]+)/,
```
Unintended change from a rebase, I'll fix it on the next push.

rafaelfranca commented on railties/test/application/per_request_fragment_cache_test.rb:
```diff
@@ -0,0 +1,59 @@
+        get '/customers/1'
+      end
+
+      assert write_not_called
```
Since it is per request should not this be false since other request was made and the cache was cleaned?

kaspth replied:
Haha, doh. Thanks :smile:

kaspth commented:
@rafaelfranca I'm really stumped on getting these tests to work :cry:

It's all coming up black magic, and I have no idea what's going on.

For some reason the initializers aren't run. However, if I added this in `cache_helper.rb`:

```ruby
class ActionView::Base
  mattr_accessor(:per_request_fragment_cache) { ActiveSupport::Cache::MemoryStore.new }
end
```

They were run! Do you have any idea why? :smile:

The next problem was that `app.config.consider_all_requests_local` was false even though I added it to the config. Can you spot if I've done it wrong?

rafaelfranca commented:
Weird. I'll debug in my machine

kaspth commented:
@dhh I finally got everything up and running here. I reread your original description of the problem and realized I'd been trying to solve the wrong thing :sweat_smile:

Anyway, the cache now captures generated digests which I think is what you want.

Also I looked at [`ActionView::Digestor`](https://github.com/rails/rails/blob/master/actionview/lib/action_view/digestor.rb) which seems to already have a cache. Can that fill your use case? :smile:

dhh commented:
Hey, hey. Yeah, all I wanted was that the cache we're already using in production for template digests to be used on a per-request basis in development. So the first lookup of a template within a request in development is computed, all subsequent lookups are not. 👍

kaspth committed:
Remove unused persistent option.

kaspth commented:
@dhh Boom, got it :+1:. Way easier once I figured out there was already a cache.

I'm closing and reopening for Travis to run.

dhh commented:
Badabing! Awesome. Thanks!
